File: rxbinding/src/main/java/com/jakewharton/rxbinding2/InitialValueObservable.java
package com.jakewharton.rxbinding2;

import io.reactivex.Observable;
import io.reactivex.Observer;

public abstract class InitialValueObservable<T> extends Observable<T> {
  @Override protected final void subscribeActual(Observer<? super T> observer) {
    subscribeListener(observer);
    observer.onNext(getInitialValue());
  }

  protected abstract void subscribeListener(Observer<? super T> observer);
  protected abstract T getInitialValue();

  public final Observable<T> skipInitialValue() {
    return new Skipped();
  }

  private final class Skipped extends Observable<T> {
    Skipped() {
    }

    @Override protected void subscribeActual(Observer<? super T> observer) {
      subscribeListener(observer);
    }
  }
}

File: rxbinding/src/main/java/com/jakewharton/rxbinding4/InitialValueObservable.kt
package com.jakewharton.rxbinding4

import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Observer

abstract class InitialValueObservable<T> : Observable<T>() {
  protected abstract val initialValue: T

  override fun subscribeActual(observer: Observer<in T>) {
    subscribeListener(observer)
    observer.onNext(initialValue)
  }

  protected abstract fun subscribeListener(observer: Observer<in T>)

  fun skipInitialValue(): Observable<T> = Skipped()

  private inner class Skipped : Observable<T>() {
    override fun subscribeActual(observer: Observer<in T>) {
      subscribeListener(observer)
    }
  }
}
